backend/src/index.js

```diff
@@ -5,140 +5,62 @@
-const statePriorities = [ 'away', 'viewing', 'commenting', 'editing', 'merging' ];
+const driver = new RedisDriver( io );
+const repository = new SessionRepository( driver );
 
 io.on( 'connection', socket => {
-	socket.session = {};
-
 	const timestamp = new Date();
 
-	socket.on( 'setUser', ( message, reply ) => {
-		const issueKey = `${ message.repoName }:${ message.issueId }`;
-		if ( !db.has( issueKey ) ) {
-			db.set( issueKey, new Map() );
-		}
-
-		if ( !socket.session.issues ) {
-			socket.session.issues = [];
-		}
-
-		if ( !socket.session.issues.includes( issueKey ) ) {
-			socket.session.issues.push( issueKey );
-		}
-
-		const issueUsers = db.get( issueKey );
-
-		if ( !issueUsers.has( message.user.login ) ) {
-			issueUsers.set( message.user.login, { sockets: [], tabs: new Map(), joinedAt: timestamp } );
-		}
-
-		const issueUser = issueUsers.get( message.user.login );
+	socket.on( 'setUser', async ( message, reply ) => {
+		const issueKey = createIssueKey( message.repoName, message.pageType, message.issueId );
 
-		if ( !issueUser.sockets.includes( socket.id ) ) {
-			socket.session.login = message.user.login;
+		const issueSession = { ...message.user, joinedAt: timestamp };
 
-			socket.join( issueKey );
+		await repository.set( issueKey, socket, issueSession );
 
-			issueUser.sockets.push( socket.id );
+		if ( !socket.issueKey ) {
+			socket.issueKey = issueKey;
 		}
 
-		issueUser.tabs.set( socket.id, message.user.state );
+		const issueSessions = await repository.getAll( issueKey );
+		socket.join( issueKey );
 
-		issueUser.user = message.user;
-
-		const users = getUsers( issueUsers );
-
-		socket.broadcast.to( issueKey ).emit( 'refresh', users );
+		const users = getUserListFromSessions( issueSessions );
 
 		reply( null, users );
-	} );
-
-	socket.on( 'removeUser', message => {
-		const issueKey = `${ message.repoName }:${ message.issueId }`;
 
-		removeUser( issueKey );
-	} );
-
-	socket.on( 'disconnect', () => {
-		for ( const issueKey of socket.session.issues || [] ) {
-			removeUser( issueKey );
-		}
-
-		delete socket.issues;
+		socket.broadcast.to( issueKey ).emit( 'refresh', users );
 	} );
 
-	function removeUser( issueKey ) {
-		const issueUsers = db.get( issueKey );
+	socket.on( 'disconnect', async () => {
+		await repository.deleteOne( socket );
+		const issueSessions = await repository.getAll( socket.issueKey );
 
-		if ( !issueUsers ) {
-			return;
-		}
-
-		const issueUser = issueUsers.get( socket.session.login );
-
-		if ( !issueUser ) {
-			return;
-		}
-
-		if ( !issueUser.sockets.includes( socket.id ) ) {
-			return;
-		}
-
-		if ( issueUser.tabs.has( socket.id ) ) {
-			issueUser.tabs.delete( socket.id );
-		}
-
-		issueUser.sockets = issueUser.sockets.filter( socketId => socketId !== socket.id );
-
-		if ( issueUser.sockets.length ) {
-			return;
-		}
+		delete issueSessions[ socket.id ];
 
-		issueUsers.delete( socket.session.login );
+		const users = getUserListFromSessions( issueSessions );
```

Comment:
I would prefer to create a method in the SessionRepository -> `getUser` instead of using helpers.  This same in `setUser` event

Reply:
Done.